dot/parachain/dispute/queues.go

```diff
@@ -0,0 +1,234 @@
+func (q ParticipationItem) Less(than btree.Item) bool {
+	other := than.(*ParticipationItem)
+
+	if q.comparator.relayParentBlockNumber == nil && other.comparator.relayParentBlockNumber == nil {
+		return bytes.Compare(q.comparator.candidateHash[:], other.comparator.candidateHash[:]) < 0
+	}
+
+	if other.comparator.relayParentBlockNumber == nil {
+		return false
+	}
+
+	if q.comparator.relayParentBlockNumber == nil {
+		return true
+	}
+
+	if isEqual := *q.comparator.relayParentBlockNumber == *other.comparator.relayParentBlockNumber; isEqual {
+		return bytes.Compare(q.comparator.candidateHash[:], other.comparator.candidateHash[:]) < 0
+	}
+
+	return *q.comparator.relayParentBlockNumber < *other.comparator.relayParentBlockNumber
+}
+
+func newParticipationItem(comparator CandidateComparator, request *ParticipationRequest) *ParticipationItem {
+	return &ParticipationItem{
+		comparator: comparator,
+		request:    request,
+	}
+}
+
+// ParticipationPriority the priority of a participation request
+type ParticipationPriority int
+
+const (
+	// ParticipationPriorityBestEffort is the lowest priority
+	ParticipationPriorityBestEffort ParticipationPriority = iota
+	// ParticipationPriorityHigh is the highest priority
+	ParticipationPriorityHigh
+)
+
+// IsPriority returns true if the priority is high
+func (p ParticipationPriority) IsPriority() bool {
+	return p == ParticipationPriorityHigh
+}
+
+var (
+	// errorBestEffortQueueFull is returned when the best effort queue is full and the request could not be processed
+	errorBestEffortQueueFull = errors.New("best effort queue is full")
+	// errorPriorityQueueFull is returned when the priority queue is full and the request could not be processed
+	errorPriorityQueueFull = errors.New("priority queue is full")
+)
+
+// Queue the dispute participation queue
+type Queue interface {
+	// Queue adds a new participation request to the queue
+	Queue(comparator CandidateComparator, request *ParticipationRequest, priority ParticipationPriority) error
+
+	// Dequeue gets the next best request for dispute participation if any.
+	Dequeue() *ParticipationItem
+
+	// PrioritiseIfPresent moves a participation request from the best effort queue to the priority queue
+	PrioritiseIfPresent(comparator CandidateComparator) error
+
+	// PopBestEffort removes the next participation request from the best effort queue
+	PopBestEffort() *ParticipationItem
+
+	// PopPriority removes the next participation request from the priority queue
+	PopPriority() *ParticipationItem
+
+	// Len returns the number of items in the specified queue
+	Len(queueType ParticipationPriority) int
+}
+
+// QueueHandler implements Queue
+// It uses two btree's to store the requests. One for best effort and one for priority.
+// The queues store participationItem's.
+// The btree is ordered by the CandidateComparator of participationItem.
+type QueueHandler struct {
+	bestEffort *btree.BTree
+	priority   *btree.BTree
```

Comment:
why aren't we using `btree.Map` from the `github.com/tidwall/btree` package? We are using it in the grandpa integration and the API is closer to the rust `BtreeMap`.

Reply:
You're right, it does seem better. I'll change it

Reply:
I took a look at the APIs and it looks like we can't use the `btree.Map` here since the key needs to be a primitive type whereas we're using `CandidateComparator` struct. I'll use `btree.BTree` instead. 

The google's btree seems API rich whereas `github.com/tidwall/btree` seems to have better performance. I'll change it

Reply:
It's done!